picasso/src/main/java/com/squareup/picasso/ExifStreamReader.java

```diff
@@ -0,0 +1,165 @@
+/*
+ * Copyright (C) 2010 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Adapted for picasso
+ */
+
+package com.squareup.picasso;
+
+import android.util.Log;
+import java.io.IOException;
+import java.io.InputStream;
+
+final class ExifStreamReader {
+  private static final String TAG = "CameraExif";
+
+  // Returns the orientation value
+  public static int getOrientation(InputStream stream) throws IOException {
+    if (stream == null) {
+      return 0;
+    }
+    MarkableInputStream markStream = new MarkableInputStream(stream);
```

Comment:
@JohnWowUs I see what you mean; I was under the impression that the `MarkableInputStream` maintains its own marker and resetting it just resets `MarkableInputStream`, but after looking at the code, I see that it also tries to reset the original `InputStream`.

However, it seems that it does not work (hence, why I'm seeing failures in decoding the images). I believe it happens when the downloader is `UrlConnectionDownloader`, since that downloader uses the `InputStream` provided by `openConnection()`, which doesn't seem resettable.

Can you also try it on your end to confirm?